pyro/contrib/minipyro.py

```diff
@@ -98,31 +125,36 @@ def __iter__(self):
         return range(self.size)
 
 
-def sample(name, fn, obs=None):
-    if not PYRO_STACK:
-        return fn()
-    msg = dict(type="sample", name=name, fn=fn, value=obs)
-    for handler in reversed(PYRO_STACK):
+def apply_stack(msg):
+    for pointer, handler in enumerate(reversed(PYRO_STACK)):
         handler.process_message(msg)
+        if msg.get("stop"):
+            break
     if msg["value"] is None:
-        msg["value"] = fn()
-    for handler in PYRO_STACK:
+        msg["value"] = msg["fn"](*msg["args"])
+    for handler in PYRO_STACK[-pointer-1:]:
```

Comment:
note: I think you could have used `reversed(enumerate(PYRO_STACK))` above, and then this would be merely `PYRO_STACK[pointer:]`?